program/src/processor/withdraw_inactive_stake.rs

```diff
@@ -17,10 +26,150 @@ use crate::instruction::accounts::{Context, WithdrawInactiveStakeAccounts};
+    );
+
+    let mut stake_data = ctx.accounts.stake.try_borrow_mut_data()?;
+    let stake = bytemuck::try_from_bytes_mut::<Stake>(&mut stake_data)
+        .map_err(|_error| ProgramError::InvalidAccountData)?;
+
+    require!(
+        stake.is_initialized(),
+        ProgramError::UninitializedAccount,
+        "stake",
+    );
+
+    let (derivation, _) =
+        find_stake_pda(&stake.validator_vote, ctx.accounts.config.key, program_id);
+
+    require!(
+        ctx.accounts.stake.key == &derivation,
+        ProgramError::InvalidSeeds,
+        "stake",
+    );
+
+    // stake authority
+    // - must be a signer
+    // - must match the authority on the stake account
+
+    require!(
+        ctx.accounts.stake_authority.is_signer,
+        ProgramError::MissingRequiredSignature,
+        "stake authority",
+    );
+
+    require!(
+        ctx.accounts.stake_authority.key == &stake.authority,
+        StakeError::InvalidAuthority,
+        "stake authority",
+    );
+
+    // vault authority
+    // - derivation must match
+
+    let bump = [config.vault_authority_bump];
+    let vault_signer = create_vault_pda(ctx.accounts.config.key, &bump, program_id)?;
+
+    require!(
+        ctx.accounts.vault_authority.key == &vault_signer,
+        StakeError::InvalidAuthority,
+        "vault authority",
+    );
+
+    let signer_seeds = get_vault_pda_signer_seeds(ctx.accounts.config.key, &bump);
+
+    // vault
+    // - must be the token account on the stake config account
+
+    require!(
+        ctx.accounts.vault.key == &config.vault,
+        StakeError::IncorrectVaultAccount,
+    );
+
+    let vault_data = ctx.accounts.vault.try_borrow_data()?;
+    // unpack to validate the mint
+    let vault = PodStateWithExtensions::<PodAccount>::unpack(&vault_data)?;
+
+    // mint
+    // - must match the stake vault mint
+
+    require!(
+        &vault.base.mint == ctx.accounts.mint.key,
+        StakeError::InvalidMint,
+        "mint"
+    );
+
+    let mint_data = ctx.accounts.mint.try_borrow_data()?;
+    let mint = PodStateWithExtensions::<PodMint>::unpack(&mint_data)?;
+    let decimals = mint.base.decimals;
+
+    // Update the config and stake account data.
+
+    require!(
+        amount <= stake.inactive_amount,
+        StakeError::NotEnoughInactivatedTokens,
+        "amount"
+    );
+
+    stake.inactive_amount = stake
+        .inactive_amount
+        .checked_sub(amount)
+        .ok_or(ProgramError::ArithmeticOverflow)?;
+
+    config.token_amount_delegated = config
+        .token_amount_delegated
+        .checked_sub(amount)
+        .ok_or(ProgramError::ArithmeticOverflow)?;
+
+    // Transfer the tokens from the vault to destination (unstakes them).
+
+    drop(vault_data);
+    drop(mint_data);
+
+    spl_token_2022::onchain::invoke_transfer_checked(
+        &spl_token_2022::ID,
+        ctx.accounts.vault.clone(),
+        ctx.accounts.mint.clone(),
+        ctx.accounts.destination_token_account.clone(),
```

Comment:
What do you think about adding a check to make sure that `vault != destination_token_account`? Not so much to protect users, but to avoid messing things up in the vault account

Reply:
Yeah, sounds good!